lib/promenade/waterdrop/statistics_subscriber.rb

```diff
@@ -0,0 +1,94 @@
+        doc "A count of the total messages delivered to Kafka"
+      end
+
+      Promenade.histogram :waterdrop_producer_delivery_attempts do
+        doc "A count of the total message deliveries attempted"
+        buckets [0, 6, 12, 18, 24, 30]
+      end
+
+      Promenade.histogram :waterdrop_producer_ack_latency_seconds do
+        doc "Delay between message being produced and Acked in seconds"
+        buckets :network
+      end
+
+      Promenade.gauge :waterdrop_async_producer_queue_size do
+        doc "Size of Kafka async producer queue"
+      end
+
+      Promenade.gauge :waterdrop_async_producer_max_queue_size do
+        doc "Max size of Kafka async producer queue"
+      end
+
+      Promenade.gauge :waterdrop_async_producer_queue_fill_ratio do
+        doc "Size of Kafka async producer queue"
+      end
+
+      Promenade.counter :waterdrop_async_producer_dropped_messages do
+        doc "Count of dropped messages"
+      end
+
+      def emitted(event)
+        statistics = event.payload[:statistics].with_indifferent_access
+        labels = get_labels(statistics)
+
+        report_root_metrics(statistics, labels)
+        report_broker_metrics(statistics[:brokers], labels)
+      end
+
+      private
+
+        def report_root_metrics(statistics, labels)
+          queue_size = statistics[:msg_cnt]
+          max_queue_size = statistics[:msg_max]
+          message_size = statistics[:msg_size]
+          delivered_messages = statistics[:txmsgs]
+          queue_fill_ratio = queue_size.to_f / max_queue_size
+
+          Promenade.metric(:waterdrop_async_producer_queue_size).set(labels, queue_size)
+          Promenade.metric(:waterdrop_async_producer_max_queue_size).set(labels, max_queue_size)
+          Promenade.metric(:waterdrop_async_producer_queue_fill_ratio).set(labels, queue_fill_ratio)
+          Promenade.metric(:waterdrop_producer_message_size).observe(labels, message_size)
+          Promenade.metric(:waterdrop_producer_delivered_messages).increment(labels, delivered_messages)
+        end
+
+        def report_broker_metrics(brokers, labels)
+          brokers.map do |broker_name, broker_values|
+            next if broker_values[:nodeid] == -1
+
+            attempts = broker_values[:txretries]
+            ack_latency = convert_microseconds_to_seconds(broker_values[:rtt][:avg])
+            broker_labels = {
+              broker_id: broker_name,
+            }
+
+            Promenade.metric(:waterdrop_producer_delivery_attempts).observe(labels.merge(broker_labels), attempts)
+            Promenade.metric(:waterdrop_producer_ack_latency_seconds).observe(labels, ack_latency)
+          end
+        end
+
+        def get_labels(statistics)
+          {
+            client: statistics[:client_id],
+          }
+        end
+
+        def convert_microseconds_to_seconds(time_in_microseconds)
+          time_in_microseconds / 1_000_000.to_f
+        end
```

Comment:
nitpick this now has turned out into a pattern: 

https://github.com/errm/promenade/blob/2dd3ed334db8987e55e9e5c216415a5c74ee2f45/lib/promenade/karafka/statistics_subscriber.rb#L75-L77


what do you think about encapsulating this behaviour in a helper Object? perhaps in a separate PR